### PR TITLE
Add option to save all traces

### PR DESCRIPTION
**Pull request by marnovandermaas**
This adds an option to QCVEngine to save all traces instead of just failures. I'm happy to implement any suggestions on how to do this in a cleaner way!

### CODE REVIEW

**Comment by PeterRugg** [comment]
Interesting. Probably a good idea (presumably for something like checking coverage?) I think this is probably the right way to do it. I'll do detailed review on Monday, but one thought: is there any reason you save out the unannotated `test` rather than `m_traceAB`, which should have comments including what the implementations said? I'm a bit fuzzy on all the types involved though, because I haven't touched this for quite a while.

**Comment by marnovandermaas** [comment]
Didn't think of trying m_traceAB. The show test looks something like:
```
# Automatically generated test case
#>QCVENGINE_TEST_V2.0
.4byte 0x00000017 # auipc x0, 0
...
# Test end
```

While showing the m_traceAB looks something like:
```
# Automatically generated test case
Just #>QCVENGINE_TEST_V2.0
((.4byte 0x99c00793 # addi x15, x0, -1636,Just Trap: False, PCWD: 0x0000000080000084, RD: 15, RWD: 0x00000000fffff99c, MA: 0x0000000000000000, MWD: 0x0000000000000000, MWM: 0b00000000, MRD: 0x0000000000000000, MRM: 0b00000000 I: 0x0000000099c00793 PRV_? XL:? (addi x15, x0, -1636)),Just Trap: False, PCWD: 0x0000000080000084, RD: 15, RWD: 0x00000000fffff99c, MA: 0x00000000fffff99c, MWD: 0x0000000000000000, MWM: 0b00000000, MRD: 0x0000000000000000, MRM: 0b00000000 I: 0x0000000099c00793 PRV_? XL:? (addi x15, x0, -1636))
...
((# Test end,Just halt token),Just halt token)
```

I think I prefer showing just the test because of readability and being able to replay the test back through TestRIG.

File: src/QuickCheckVEngine/Main.hs [hunk 1]
```diff
@@ -327,15 +332,15 @@ main = withSocketsDo $ do
...
         where onDeath test = do putStrLn "Failure rerunning test"
                                 askAndSave sourceFile (show test) Nothing testTrans
               onTrace trace = askAndSave sourceFile (showAnnotatedTrace (isNothing m_implB) archDesc trace) (Just trace) testTrans
   let checkTrapAndSave sourceFile test = saveOnFail sourceFile test (check_mcause_on_trap :: Test TestResult -> Test TestResult)
   let checkResult = if optVerbosity flags > 1 then verboseCheckWithResult else quickCheckWithResult
   let checkGen gen remainingTests =
         checkResult (Args Nothing remainingTests 1 (testLen flags) (optVerbosity flags > 0) (if optShrink flags then 1000 else 0))
-                    (prop implA m_implB alive (checkTrapAndSave Nothing) archDesc (timeoutDelay flags) (optVerbosity flags) (optIgnoreAsserts flags) (optStrict flags) gen)
+                    (prop implA m_implB alive (checkTrapAndSave Nothing) archDesc (timeoutDelay flags) (optVerbosity flags) (if (optSaveAll flags) then (saveDir flags) else Nothing) (optIgnoreAsserts flags) (optStrict flags) gen)
```
Comment (PeterRugg) [inline src/QuickCheckVEngine/Main.hs:343]: Slight issue: users might be surprised if they passed "save-all", forgot to pass a directory, and lose all the tests. I wonder if it's convenient to give an error early on in that case? We could also have separate options "failures-dir" and "tests-dir" that the user can set separately.

**Review by PeterRugg — APPROVED** [review]
Happy, except for potential usability improvement if it's convenient.

File: src/QuickCheckVEngine/Main.hs (hunk 1 above)
Comment (marnovandermaas) [reply src/QuickCheckVEngine/Main.hs:343]: Yeah, locally I had put that check into TestRIG run script. I might merge this PR and then see if I can get the failure and test directories separated.

**Pull request merged**